extmod/btstack/modbluetooth_btstack.c

```diff
@@ -751,19 +938,40 @@ int mp_bluetooth_gattc_read(uint16_t conn_handle, uint16_t value_handle) {
 int mp_bluetooth_gattc_write(uint16_t conn_handle, uint16_t value_handle, const uint8_t *value, size_t *value_len, unsigned int mode) {
     DEBUG_EVENT_printf("mp_bluetooth_gattc_write\n");
 
-    // TODO the below gatt_client functions do not copy the data and require it to be valid
-    // until the write is done, so there should be some kind of buffering done here.
+    // We should be distinguishing between gatt_client_write_value_of_characteristic vs
+    // gatt_client_write_characteristic_descriptor_using_descriptor_handle.
+    // However both are implemented using send_gatt_write_attribute_value_request under the hood,
+    // and we get the exact same event to the packet handler.
+    // Same story for the "without response" version.
+
+    int err;
 
     if (mode == MP_BLUETOOTH_WRITE_MODE_NO_RESPONSE) {
-        // TODO need to call gatt_client_request_can_write_without_response_event then do
-        // the actual write on the callback from that.
-        return btstack_error_to_errno(gatt_client_write_value_of_characteristic_without_response(conn_handle, value_handle, *value_len, (uint8_t *)value));
-    }
-    if (mode == MP_BLUETOOTH_WRITE_MODE_WITH_RESPONSE) {
-        return btstack_error_to_errno(gatt_client_write_value_of_characteristic(&btstack_packet_handler_write_with_response, conn_handle, value_handle, *value_len, (uint8_t *)value));
+        // If possible, this will send immediately, copying the buffer directly to the ACL buffer.
+        err = gatt_client_write_value_of_characteristic_without_response(conn_handle, value_handle, *value_len, (uint8_t *)value);
+        if (err == GATT_CLIENT_BUSY) {
+            DEBUG_EVENT_printf("mp_bluetooth_gattc_write: client busy\n");
+            // Can't send right now, need to take a copy of the buffer and add it to the queue.
+            btstack_enqueue_pending_operation(MP_BLUETOOTH_BTSTACK_PENDING_WRITE_NO_RESPONSE, conn_handle, value_handle, value, *value_len);
+            // Notify when this conn_handle can write.
+            err = gatt_client_request_can_write_without_response_event(&btstack_packet_handler_generic, conn_handle);
```

Comment:
if this returns an error, shouldn't `btstack_finish_pending_operation()` be called to unref the buffer?